memchain/src/tests.rs

```diff
@@ -1,64 +1,76 @@
 #![cfg(test)]
 
+use blockchain_traits::PendingTransaction;
+
 use crate::*;
 
 const ADDR_1: Address = Address([1u8; 20]);
 const ADDR_2: Address = Address([2u8; 20]);
 
+const BASE_GAS: u64 = 2100;
+
 fn giga(num: u64) -> u64 {
     num * 1_000_000_000
 }
 
-extern "C" fn nop_main(_bc: *const *mut dyn Blockchain<Address = Address>) -> u16 {
+extern "C" fn nop_main(
+    _ptx: *const *mut dyn PendingTransaction<Address = Address, AccountMeta = AccountMeta>,
+) -> u16 {
     0
 }
 
-extern "C" fn simple_main(bc: *const *mut dyn Blockchain<Address = Address>) -> u16 {
-    let bc = unsafe { &mut **bc };
+extern "C" fn simple_main(
+    ptx: *const *mut dyn PendingTransaction<Address = Address, AccountMeta = AccountMeta>,
+) -> u16 {
+    let ptx = unsafe { &mut **ptx };
 
-    assert_eq!(bc.sender(), &ADDR_2);
+    assert_eq!(ptx.sender(), &ADDR_2);
 
-    bc.emit(vec![[42u8; 32]], vec![0u8; 3]);
+    ptx.emit(vec![[42u8; 32].as_ref()].as_slice(), &[0u8; 3]);
 
-    let mut rv = bc.fetch_input();
+    let mut rv = ptx.input().to_vec();
     rv.push(4);
-    bc.ret(rv);
+    ptx.ret(&rv);
 
     0
 }
 
-extern "C" fn fail_main(bc: *const *mut dyn Blockchain<Address = Address>) -> u16 {
-    let bc = unsafe { &mut **bc };
-    bc.err(r"¯\_(ツ)_/¯".as_bytes().to_vec());
+extern "C" fn fail_main(
+    ptx: *const *mut dyn PendingTransaction<Address = Address, AccountMeta = AccountMeta>,
+) -> u16 {
+    let ptx = unsafe { &mut **ptx };
+    ptx.err(r"¯\_(ツ)_/¯".as_bytes());
```

Comment:
useful error. any way to have this take a useful error message?

Reply:
does this not demonstrate the interface?